### ISSUE TITLE
Should timezone awareness be a part of datetime's static type?

### ISSUE THREAD

**Issue by ronawho** [chapel-lang/chapel#9941]
In the current DateTime implementation, timezone awareness is determined by whether datetime's tzinfo is nil or not. Operations between mixed datetimes are not allowed, which means that either both datetimes must have timezones or neither can.

In today's implementation this leads to lots of runtime checks for mixed timezone awareness, and mismatches currently lead to halts: https://github.com/chapel-lang/chapel/blob/805996b81270af5e2b3e96cc8d8dfaf8ab668df3/modules/standard/DateTime.chpl#L1301-L1319

In looking to convert DateTime over to error-handling, it seems like a shame to have to make these operators throwing, since that seems burdensome for users. e.g. for something like:

```chpl
use DateTime;

var t1 = datetime.now();
var t2 = datetime.now();
if t1 <= t2 then writeln(t1);
```

Ignoring errors would look something like: 

```chpl
use DateTime;

var t1 = datetime.now();
var t2 = datetime.now();
if (try! t1 <= t2) then writeln(t1); // try! as part of the condition looks weird to me
```

And if you did want to handle errors would look something like:

```chpl
use DateTime;

var t1 = datetime.now();
var t2 = datetime.now();
var cmp: bool;
try {
  cmp = t1 <= t2;
} catch {
  // handle error
}
if cmp then writeln(t1);
```

Mismatched timezones feels like a type error to me. This made me wonder if we should just make timezone awareness be part of the type, which would allow us to perform these checks at compile time.

**Comment by riftEmber** [comment]
In https://github.com/Cray/chapel-private/issues/5170 we decided to do this for 2.0, and implementation is in progress at https://github.com/chapel-lang/chapel/pull/23010.

**Comment by jeremiah-corrado** [comment]
In attempting to implement this change, I'm realizing there may need to be a few additional API changes to the `time` and `dateTime` types to smoothly transition from runtime timezone-awareness to compile time timezone-awareness.

Under the current design, some `time` and `dateTime` factory methods and initializers take nilable `Timezone` arguments, where a `nil` value results in a timezone-unaware (or naiive) value, and a non-nil argument results in a timezone-aware value. This does not mix well with the new design where a `param tzAware: bool` field would designate the timezone awareness of the value.

### Specifics of Problem:

Consider the `replace` interface, which can be used to create a new `time` or `dateTime` that inherits the fields from an existing value, except with some of its values replaced. Example:

```chapel
var dt2 = dt1.replace(day=3);
var t2 = t1.replace(hour=1, minute=5);
```

These methods also allow creation of a `time`/`dateTime` where the timezone is removed or added. Example:
```chapel
var dtNoTZ = dt1.replace(tz=nil);  // remove the timezone
var dt2 = dtNoTZ.replace(tz=myTimeZone); // add a different timezone
```

This is problematic in a framework where timezone-awareness is a static property of a `time` or `dateTime`, because the return type of `replace` cannot be determined at compile time. I.e., the method signature for `replace` would need to look something like the following, which is obviously not possible:
```chapel
proc dateTime.replace(year=-1, month=-1, day=-1,
                        hour=-1, minute=-1, second=-1, microsecond=-1,
                        in tz: shared Timezone? = this.timezone): dateTime(tzAware = (tz != nil)) { ... }
```

The same problem arises with the initializers on `time` and `dateTime`. However, separate overloads can be added for creating aware/naive values with distinct return types.

### Recommendation:

To get around this, I would suggest adding some new methods to `time`/`dateTime` to be used instead of `replace(tz=nil)` or `replace(tz=otherTZ)`. The signatures might look like:
```chapel
proc dateTime.withTZ(in tz: shared Timezone): dateTime(tzAware=true);
proc dateTime.stripTZ(): dateTime(tzAware=false);
```

Given the time crunch with the next release, I'm going to hold off on this change until there is time for more design discussion. See my partially completed implementation here: https://github.com/chapel-lang/chapel/pull/23359

**Comment by riftEmber** [comment]
@jeremiah-corrado I figured for `replace` methods we would have an overload taking a Timezone and another without, as well as use `where`s to have different return types based on whether the receiver was aware or naive. So return types could be:

- aware `dateTime`, no-tz replace overload: aware `dateTime`
- aware `dateTime`, with-tz replace overload: aware `dateTime`
- naive `dateTime`, no-tz replace overload: naive `dateTime`
- naive `dateTime`, with-tz replace overload: aware `dateTime`

This isn't particularly satisfying though, and would also leave us without a way to make an aware dateTime naive. I would be on board with removing all timezone functionality (and formals) from `replace` methods, and adding a `withTz` and `stripTz` to allow a user to change those components.

**Comment by jeremiah-corrado** [comment]
> This isn't particularly satisfying though, and would also leave us without a way to make an aware dateTime naive.

I agree, I did something similar with where-clauses [here](https://github.com/chapel-lang/chapel/pull/23359/files#diff-97b289e5fce1179b949e8d692a8ba925a5cd03db09afe338764e26c886de1023), but still ended up needing a `stripTZ` method unfortunately.
